### ISSUE TITLE
Adding Available Applications to any media device throws "Applications disabled" error

### ISSUE THREAD

**Issue by rockstar2020** [mikejac/node-red-contrib-google-smarthome#334]
Hello,
I have several media devices which I have setup over past couple of years. and all are triggering automations with no issues.
Today I wanted to add a new media device (ie Streaming Box) and control few automations using the applications  installed on this media device but when I trigger adding the applications with the usual code as below I now get an error message of "Applications disabled".

```json
[
    {
        "key": "Sleep",
        "names": [
            "Sleep",
            "My Sleep"
        ]
    },
    {
        "key": "New Pop",
        "names": [
            "New Pop"
        ]
    },
    {
        "key": "New Rock",
        "names": [
            "New Rock"
        ]
    },
    {
        "key": "Lounge",
        "names": [
            "Lounge",
            "My Lounge"
        ]
    }
]
```

After getting this error message I can see the JSON file in .node-red folder is created but it's empty.
Even if I copy exactly another application's JSON file which is working perfectly fine I still cannot lauch any application on the newly added media device.
Any thoughts what potentially has caused this issue?

Many thanks in advance

**Comment by Caprico85** [comment]
Hi,

I never used streaming boxes or the AppSelector trait (or any of the traits with file-based settings) and I didn't wrote that piece of the code. So I may not be of much help here.

Judging by the code, the message "Applications disabled" should only ever happen if a device receives a message with the topic "availableapplications" and the AppSelector trait is disabled.

This brings us to the question: Is the App Selector trait enabled? Which device type did you select?

![image](https://user-images.githubusercontent.com/2081806/197613050-e953cfb2-03b8-4c22-a8b0-fee76e9dc116.png)

For the error message "Applications disabled", it should not matter if the applications file is readable or valid. The error message happens before the module even tries to access the file.

I'm currently trying to reproduce your problem. So far, Im able to send in new applications via an "availableapplications" message and the apps are successfully written to the file. But I'm not able to run any of the apps. Google always say it can't find an app with that name. I'l look into that in the next days.

Nevertheless, some other ideas coming to my mind:
- What happens if you set the apps in the device config itself?
  ![image](https://user-images.githubusercontent.com/2081806/197623098-455f7290-7696-4514-a403-4c4f6d777d84.png)
- File permissions on the folder and file are correct? (I know, permissions should not matter in your case as the error message happens earlier ... but you never know)

**Comment by rockstar2020** [comment]
Hi @Caprico85, 
Thank you for your quick response to issues reported as always.
I also though may be the issue would be with file permissions but that's not the case because as I mentioned earlier I did copy an existing working application list to a file that this new media device is pointing to.
After I manually added the JSON file I don't get the "Applications Disabled" error but still when I ask Google to launch/open any of the registered applications Google says that it can't find the said app.
Additionally as you know, there's no way to select or deselect App Selector so I thought adding a file name or typing the JSON directly into the field will enable it automatically.
Anyways, I have had media device added to my automation setup for quite sometime and never had an issue with adding Apps or Channels to it. This is the first time I've come across this error message.

Thank you for your help!


Cheers

**Comment by FireWizard52** [comment]
Hello @rockstar2020 

I do not use a media device, but I use a Device type "Security System". See#231
This device requires also a JSON configuration.
In the past I had a JSON file configured, but this file has disappeared. Probably caused by some update.
I have also the configuration configured in the node.
This has always worked for me.

Perhaps, if you have solved it, we have to add the configuration of such a device to to the examples folder.
We can do that with the Security System, as well

@Caprico85 Wat is your opinion?

**Comment by rockstar2020** [comment]
Hey guys,
Just wanted to confirm that the new media device I added works properly if I put the applications keys/names as JSON right into the node but surely doesn't work if it is saved in node-red folder as a JSON file like it used to work before.


Cheers

**Comment by Caprico85** [comment]
It seems Google has changed something. When I tested during the last two days, Google always sent the application key in lowercase, even if it was defined uppercase. That's why running an app failed, because "Sleep" != "sleep". When I tested today, the name was sent as defined ("Sleep") and the apps on my test device suddenly started to work. Application names are still sent wrong though (defined "New Rock", sent "new Rock"). This is why I'm currently working on making the string comparison case insensitive.

When I'm done with this, I'll have a look at the JSON files again to see if I can reproduce the problems. 


> Perhaps, if you have solved it, we have to add the configuration of such a device to the examples folder.
> We can do that with the Security System, as well

I totally agree. I'd love to have more examples, especially for the more complex devices like streaming boxes or security systems.

**Comment by rockstar2020** [comment]
@Caprico85, although as you said file permission might not be the root cause of the error "Applications Disabled" I still did check and all owners of the files and folders and it is "pi" user which is the way it should be.
As mentioned before when I push the applications list via Inject Node the JSON file gets created but it's empty. But here is how I could get the Inject Node push to work:
1. Add the JSON contents in the Media Device Node and deploy.
2. Copy an existing application JSON file wihtin .node-red folder and rename it (ie test-apps.json).
3. Change the App Selector in Device Node to point to a file I created manually (test-apps.json) and then Deploy.
4. After this if I trigger the Inject Node I don't get any error and the file gets updated with the application list I have in Inject Node.

I think there's something wrong in the code.

FYI I'm using the latest version 0.3.13.

Cheers

**Comment by Caprico85** [comment]
I'm not sure anymore if it's a problem with file permissions. Permission problems or files not found would result in an error message in the log and in Node-RED's debug panel. I guess you don't have that message?

![image](https://user-images.githubusercontent.com/2081806/198971278-daaaab4b-2b1c-49d1-9313-24177cba637c.png)


I found a way to reproduce the "Applications disabled" message. If there is an empty app list (`{}` or completely empty) set either as file or JSON, the AppSelector trait is disabled internally, even if it is enabled in the config. If you try to inject a new app list later, the device just responds with "Applications disabled" as the trait is still disabled.

I think having an empty initial list and injecting the app list later is a legitimate use case. Therefore I'm working on fixing that. 

I'm not sure if this is related to your case as your app list isn't empty. Anyway, I'll report back when I'm done. Maybe your problem is fixed with this too.

**Comment by ckhmer1** [comment]
Hi, 
I think this is the root cause.
You should remove the line that disable the trait and add a check for the report state, if the applications list cannot be empty.
Those changes should be applied also for other trait.

**Comment by Caprico85** [comment]
I released v0.3.14. This version does not silently disable traits anymore which should fix the "Applications disabled" error. It is now possible to start wirh an empty file or JSON string and inject a new list with an availableapplication message later.

**Comment by rockstar2020** [comment]
Great, thanks @Caprico85 
I'll give it a try.
Cheers
